README
NAME
    Catalyst::Plugin::FormValidator::Simple::Auto - Smart validation with
    FormValidator::Simple

SYNOPSIS
    Loading:

        use Catalyst qw/
          ConfigLoader
          FormValidator::Simple
          FormValidator::Simple::Auto
          /;

    myapp.yml:

        validator:
          profiles: __path_to(profiles.yml)__
          messages: __path_to(messages.yml)__
          message_format: <span class="error">%s</span>
          # and other FormValidator::Simple config

    profiles.yml:

        # profiles.yml
        action1:
          param1:
            - NOT_BLANK
            - ASCII
            - [ 'LENGTH', 4, 10 ]
          param2:
            - NOT_BLANK

    Then, using in your action:

        sub action1 : Global {
            my ($self, $c) = @_;
        
        # $c->form($profile) already executed!
            unless ($c->form->has_error) {
                ...
            }
        }

DESCRIPTION
    This plugin provide auto validation to Plugin::FormValidator::Simple.

    You can define validation profiles into config or YAML file, and no
    longer have to write it in actions.

BUNDLING ERROR MESSAGES
    With 0.08 or earlier version, you need define validator profiles and
    error messages in separately.

    So, you had to write two settings like this:

        # profiles.yml
        action1:
          param1:
            - NOT_BLANK
        
    # messages.yml
        action1:
          param1:
            NOT_BLANK: param1 is required!

    It's bothered!

    Since 0.09, you can place error messages in profiles config.

    Above two configs is equals to:

        # profiles.yml
        action1:
          param1:
            - rule: NOT_BLANK
              message: param1 is required!

EXTENDED METHODS
  setup
  prepare
  forward
ORIGINAL IDEA
    Daisuke Maki <dmaki@cpan.org>

AUTHOR
    Daisuke Murase <typester@cpan.org>

COPYRIGHT
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl module Catalyst::Plugin::FormValidator::Simple::Auto

0.17 2008-04-09T18:57:31+09:00
     - use local when set $YAML::UseAliases

0.16 2008-04-09T16:24:45+09:00
     - support yaml aliases (Reported by Nakane)

0.15 2007-04-10T10:56:52+09:00
     - support bundle message that has no validation rule

0.14 2007-04-08T02:46:50+09:00
     - don't delete not only profile action but also form object, too

0.13 2007-04-08T02:03:57+09:00
     - don't delete profile if it's first one

0.12 2007-02-22T17:14:20+09:00
	 - Added nest forward validation (Thanks to Daniel Burke)
	 - Added form_messages method

0.11 2007-01-02T02:56:00+09:00
     - Fixed cannot handle error_messages when validation rule is array ref

0.10 2006-12-26T09:52:38+09:00
     - Fixed wrong yaml format in pod
     - Make yaml loading looser

0.09 2006-12-23T08:58:04+09:00
     - Use warn instead of error, when no validator profile set
     - Switched to Module::Install
     - Added validation error message bundling feature

0.08 2006-09-09T15:16:29+09:00
     - removed $c->path_to when loading yaml file

0.07 2006-07-25T13:05:01+09:00
     - added validator_profile accessor to store current profile
     - added some basic tests

0.06 2006-07-24T02:44:21+09:00
     - fixed forward action tracker bug when action is relative calls

0.05 2006-07-24T01:48:38+09:00
     - added forward tracking

0.04 2006-07-07T11:56:59+09:00
     - fixed yaml load error with latest YAML.pm

0.03 2006-06-14T11:30:20+09:00
     - fixed bug for action name (reported by tokuhirom)

0.02 2006-06-14T02:08:49+09:00
     - added missing prereq modules (no code changes)

0.01 2006-06-14T01:44:54+09:00
     - original version;


lib/Catalyst/Plugin/FormValidator/Simple/Auto.pm
package Catalyst::Plugin::FormValidator::Simple::Auto;
use strict;
use warnings;
use base qw/Class::Accessor::Fast/;

use Catalyst::Exception;
use UNIVERSAL::isa;
use YAML;
use FormValidator::Simple;

our $VERSION = '0.17';

__PACKAGE__->mk_accessors(qw/validator_profile/);

=head1 NAME

Catalyst::Plugin::FormValidator::Simple::Auto - Smart validation with FormValidator::Simple

=head1 SYNOPSIS

Loading:

    use Catalyst qw/
      ConfigLoader
      FormValidator::Simple
      FormValidator::Simple::Auto
      /;

myapp.yml:

    validator:
      profiles: __path_to(profiles.yml)__
      messages: __path_to(messages.yml)__
      message_format: <span class="error">%s</span>
      # and other FormValidator::Simple config

profiles.yml:

    # profiles.yml
    action1:
      param1:
        - NOT_BLANK
        - ASCII
        - [ 'LENGTH', 4, 10 ]
      param2:
        - NOT_BLANK

Then, using in your action:

    sub action1 : Global {
        my ($self, $c) = @_;
    
        # $c->form($profile) already executed!
        unless ($c->form->has_error) {
            ...
        }
    }

=head1 DESCRIPTION

This plugin provide auto validation to Plugin::FormValidator::Simple.

You can define validation profiles into config or YAML file, and no longer have to write it in actions.

=head1 BUNDLING ERROR MESSAGES

With 0.08 or earlier version, you need define validator profiles and error messages in separately.

So, you had to write two settings like this:

    # profiles.yml
    action1:
      param1:
        - NOT_BLANK
    
    # messages.yml
    action1:
      param1:
        NOT_BLANK: param1 is required!

It's bothered!

Since 0.09, you can place error messages in profiles config.

Above two configs is equals to:

    # profiles.yml
    action1:
      param1:
        - rule: NOT_BLANK
          message: param1 is required!

=head1 METHODS

=head2 validator_profile

An accessor for current profile name

=head2 form_messages

=head2 form_messages( 'key' );

Return error messages about current validator profile.

If the key isn't presented, return all messages as hash.

=cut

sub form_messages {
    my ($c, $key) = @_;
    my $messages = $c->form->field_messages( $c->validator_profile );

    $key ? $messages->{$key} : $messages;
}

=head1 EXTENDED METHODS

=head2 setup

=cut

sub setup {
    my $c = shift;
    my $config = $c->config->{validator};

    Catalyst::Exception->throw( message =>
          __PACKAGE__ . qq/: You need to load "Catalyst::Plugin::FormValidator::Simple"/ )
      unless $c->isa('Catalyst::Plugin::FormValidator::Simple');

    $c->log->warn( __PACKAGE__ . qq/: You must set validator profiles/ )
      unless $config->{profiles};

    if ( $config->{profiles} and ref $config->{profiles} ne 'HASH' ) {
        my $profiles = eval {
            no warnings 'once';
            local $YAML::UseAliases = 0;
            YAML::Load( YAML::Dump( YAML::LoadFile( $config->{profiles} ) ) ); # XXX: remove yaml aliases
        };
        Catalyst::Exception->throw( message => __PACKAGE__ . qq/: $@/ ) if $@;

        $config->{profiles} = $profiles;
    }

    my $messages;
    my $profiles = $config->{profiles};
    for my $action ( keys %{ $profiles || {} } ) {
        my $profile = $profiles->{$action} || {};

        for my $param ( keys %$profile ) {
            my $rules = $profile->{$param} || [];

            my $i = 0;
            for my $rule (@$rules) {
                if ( ref $rule eq 'HASH' and defined $rule->{rule} ) {
                    my $rule_name = ref $rule->{rule} eq 'ARRAY' ? $rule->{rule}[0] : $rule->{rule};
                    $messages->{$action}{$param} ||= {};
                    $messages->{$action}{$param}{ $rule_name } = $rule->{message} if defined $rule->{message};
                    $rule = $rule->{rule};
                }
                elsif (ref $rule eq 'HASH' and defined $rule->{self_rule} ) {
                    $messages->{$action}{$param} ||= {};
                    $messages->{$action}{$param}{ $rule->{self_rule} } = $rule->{message} if defined $rule->{message};
                    delete $rules->[$i];
                }
                $i++;
            }
        }
    }

    # XXX: replace messages. ponder about hash merging
    $config->{messages} = $messages if $messages;

    # fix plugin order
    {
        no strict 'refs';
        my $pkg = __PACKAGE__;
        @{"$c\::ISA"} = ( $pkg, grep !/^$pkg$/, @{"$c\::ISA"} );
    }

    $c->NEXT::setup(@_);
}

=head2 prepare

=cut

sub prepare {
    my $c = shift->NEXT::prepare(@_);

    if ( my $profile = $c->config->{validator}{profiles}{ $c->action->reverse } ) {
        $c->validator_profile( $c->action->reverse );
        $c->form(%$profile);
    }

    $c
}

=head2 forward

=cut

sub forward {
    my $c = shift;
    my $action = $c->dispatcher->_invoke_as_path($c, @_);

    no warnings 'once';
    local $NEXT::NEXT{ $c, 'forward' };

    my $res;
    if ( my $profile = $c->config->{validator}{profiles}{ $action } ) {
        # first time validation
        if (not $c->validator_profile) {
            $c->{validator_profile} = $action;

            $c->form(%$profile);
            $res = $c->NEXT::forward(@_);
        }
        else {
            # don't override validator stuffs when not first time validation
            local $c->{validator} = FormValidator::Simple->new;
            local $c->{validator_profile} = $action;

            $c->form(%$profile);
            $res = $c->NEXT::forward(@_);
        }
    }
    else {
        $res = $c->NEXT::forward(@_);
    }

    $res;
}

=head1 ORIGINAL IDEA

Daisuke Maki <dmaki@cpan.org>

=head1 AUTHOR

Daisuke Murase <typester@cpan.org>

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut

1;
